src/linker/Linker/UnconditionalSuppressMessageAttributeState.cs

```diff
@@ -0,0 +1,244 @@
+		}
+
+		public void AddLocalSuppression (CustomAttribute ca, MetadataToken mdToken)
+		{
+			SuppressMessageInfo info;
+			if (!TryDecodeSuppressMessageAttributeData (ca, out info)) {
+				return;
+			}
+
+			if (_localSuppressionsByMdToken.ContainsKey (mdToken)) {
+				_localSuppressionsByMdToken[mdToken].Add (info.Id, info);
+			} else {
+				_localSuppressionsByMdToken[mdToken] = new Dictionary<string, SuppressMessageInfo> { { info.Id, info } }.ToImmutableDictionary ();
+			}
+
+			HasLocalSuppressions = true;
+		}
+
+		public bool IsSuppressed (string id, MessageOrigin warningOrigin, out SuppressMessageInfo info)
+		{
+			info = default;
+
+			if (HasGlobalSuppressions && IsGloballySuppressed (id, out info)) {
+				return true;
+			}
+
+			if (HasSuppressions && warningOrigin.MdTokenProvider != null) {
+				IMetadataTokenProvider mdTokenProvider = warningOrigin.MdTokenProvider;
+				do {
+					if (IsLocallySuppressed (id, mdTokenProvider.MetadataToken, out info)) {
+						return true;
+					}
+
+					if (mdTokenProvider is FieldDefinition ||
+						mdTokenProvider is PropertyDefinition ||
+						mdTokenProvider is EventDefinition ||
+						mdTokenProvider is MethodDefinition) {
+						mdTokenProvider = (mdTokenProvider as IMemberDefinition).DeclaringType;
+					} else if (mdTokenProvider is TypeDefinition type) {
+						if (type.IsNested) {
+							_ = type.BaseType.Resolve ();
+						}
+
+						mdTokenProvider = type.Module;
+					} else if (mdTokenProvider is ModuleDefinition module) {
+						mdTokenProvider = null;
+					}
+				} while (mdTokenProvider != null);
+			}
+
+			return false;
+		}
+
+		private bool TryDecodeSuppressMessageAttributeData (CustomAttribute attribute, out SuppressMessageInfo info)
+		{
+			info = default;
+
+			// We need at least the Category and Id to decode the warning to suppress.
+			// The only UnconditionalSuppressMessageAttribute constructor requires those two parameters.
+			if (attribute.ConstructorArguments.Count < 2) {
+				return false;
+			}
+
+			// Ignore the category parameter because it does not identify the warning
+			// and category information can be obtained from warnings themselves.
+			info.Id = attribute.ConstructorArguments[1].Value as string;
+			if (info.Id == null) {
+				return false;
+			}
+
+			// Allow an optional human-readable descriptive name on the end of an Id.
+			// See http://msdn.microsoft.com/en-us/library/ms244717.aspx
+			var separatorIndex = info.Id.IndexOf (':');
+			if (separatorIndex != -1) {
+				info.Id = info.Id.Remove (separatorIndex);
+			}
+
+			if (attribute.HasProperties) {
+				foreach (var p in attribute.Properties) {
+					switch (p.Name) {
+					case "Scope":
+						info.Scope = p.Argument.Value as string;
+						break;
+					case "Target":
+						info.Target = p.Argument.Value as string;
+						break;
+					case "MessageId":
+						info.MessageId = p.Argument.Value as string;
+						break;
```

Comment:
I think there should be warning if any of the values are in incorrect format

Reply:
I agree, although because IL doesn't keep information about the source line where a custom attribute is, I don't see a good way to report warnings on this scenario where the CA can be put anywhere.

Reply:
I checked how Roslyn reacts and it basically ignores invalid values. I think we should probably do the same. Maybe check that the types match, but otherwise don't validate the actual string values.